authors/apps/articles/views.py

```diff
@@ -506,3 +507,88 @@ def get(self, request):
             "favorites": favorite_articles
         }
         return Response(data=favorites, status=status.HTTP_200_OK)
+
+
+class BookmarkView(generics.CreateAPIView):
+    """This class creates a new bookmark"""
+    permission_classes = (IsAuthenticated,)
+    serializer_class = BookmarkSerializer
+
+    def post(self, request, slug):
+        """Creates a bookmark"""
+        data = request.data
+        article = Article.objects.filter(
+            slug=slug, published=True, activated=True).first()
+        if article is None:
+            not_found = {
+                "detail": "This article has not been found."
+            }
+            return Response(data=not_found, status=status.HTTP_404_NOT_FOUND)
+
+        bookmark = Bookmark.objects.filter(
+            user_id=request.user.pk, article_id=article.id).first()
+        if bookmark is not None:
+            favorite_found = {
+                "detail": "Article already bookmarked."
+            }
+            return Response(data=favorite_found,
+                            status=status.HTTP_400_BAD_REQUEST)
+
+        data['article_id'] = article.id
+        data['user_id'] = request.user.pk
+        serializer = self.get_serializer(data=data)
+        serializer.is_valid(raise_exception=True)
+        self.perform_create(serializer)
+        data['detail'] = 'Article added to bookmarks.'
+        return Response(data, status=status.HTTP_201_CREATED)
+
+    def delete(self, request, *args, **kwargs):
+        article = Article.objects.filter(
+            slug=kwargs['slug'], published=True, activated=True).first()
+        if article is None:
+            not_found = {
+                "detail": "This article has not been found."
+            }
```

Comment:
@ElMonstro, similar blocks of code is found within the same file (for example line 524), consider refactoring.

Reply:
I will work on that @abulojoshua1

Reply:
I have made some changes @abulojoshua1, see if they are satisfactory